meta/retention_policy_enforcer.go

```diff
@@ -1,65 +1,48 @@
 package meta
 
-type RetentionPolicyEnforcer struct {
-}
+import (
+	"time"
+)
 
-/*
-// StartRetentionPolicyEnforcement launches retention policy enforcement.
-func (s *Server) StartRetentionPolicyEnforcement(checkInterval time.Duration) error {
-	if checkInterval == 0 {
-		return fmt.Errorf("retention policy check interval must be non-zero")
+// RetentionPolicyEnforcer enforces retention policies.
+type RetentionPolicyEnforcer struct {
+	Store interface {
+		Databases() (dis []DatabaseInfo, err error)
+		DeleteShardGroup(database, policy string, id uint64) error
 	}
-	rpDone := make(chan struct{}, 0)
-	s.rpDone = rpDone
-	go func() {
-		for {
-			select {
-			case <-rpDone:
-				return
-			case <-time.After(checkInterval):
-				s.EnforceRetentionPolicies()
-			}
-		}
-	}()
-	return nil
 }
 
-// EnforceRetentionPolicies ensures that data that is aging-out due to retention policies
-// is removed from the server.
-func (s *Server) EnforceRetentionPolicies() {
-	log.Println("retention policy enforcement check commencing")
+// NewRetentionPolicyEnforcer returns a retention policy enforcer using the given store.
+func NewRetentionPolicyEnforcer(store *Store) *RetentionPolicyEnforcer {
+	return &RetentionPolicyEnforcer{Store: store}
+}
 
-	type group struct {
-		Database  string
-		Retention string
-		ID        uint64
+// Enforce removes shard groups that represent data older than the retention policies
+// allow. It then returns a slice of the shard IDs that must be deleted.
+func (r *RetentionPolicyEnforcer) Enforce() ([]uint64, error) {
+	dis, err := r.Store.Databases()
+	if err != nil {
+		return nil, err
 	}
 
-	var groups []group
-	// Only keep the lock while walking the shard groups, so the lock is not held while
-	// any deletions take place across the cluster.
-	func() {
-		s.mu.RLock()
-		defer s.mu.RUnlock()
+	// Check all shard groups.
+	var ids []uint64
+	for _, di := range dis {
+		for _, rp := range di.RetentionPolicies {
+			for _, g := range rp.ShardGroups {
+				if rp.Duration != 0 && g.EndTime.Add(rp.Duration).Before(time.Now().UTC()) {
+					for _, sh := range g.Shards {
+						ids = append(ids, sh.ID)
+					}
```

Comment:
If we invert this looping logic to:
1. Delete all expired shard groups
2. Loop over each shard in the `tsdb.Store` and ask the metastore whether the shard ID is still valid

Then shards would get cleaned up when shard groups expire, retention policies are removed as well as databases are deleted and we wouldn't need to return the shard IDs.

Basically suggesting more of a garbage collection approach to removing shards.